### ISSUE TITLE
Implement RTMDet to Perception Pipeline

### ISSUE THREAD

**Issue by StepTurtle** [autowarefoundation/autoware.universe#7235]
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I've agreed with the maintainers that I can plan this task.

### Description

We plan to add the  [RTMDet](https://github.com/open-mmlab/mmdetection/tree/3.x/configs/rtmdet#rtmdet-an-empirical-study-of-designing-real-time-object-detectors) model in addition to the existing YOLOX model in Autoware Universe. While the YOLOX model is successful in the bounding box task, its instance segmentation layer is weak. We aim to improve instance segmentation results by adding RTMDet model.

![208044554-1e8de6b5-48d8-44e4-a7b5-75076c7ebb71](https://github.com/autowarefoundation/autoware.universe/assets/76053179/155f45e9-5003-4e26-b5cd-6f67c665f4c7)

![208070055-7233a3d8-955f-486a-82da-b714b3c3bbd6](https://github.com/autowarefoundation/autoware.universe/assets/76053179/fde70467-2c53-4c83-a9c2-b24c14130e02)

### Purpose

Our goal is to enhance the lidar image fusion pipeline by adding the RTMDet model to Autoware for image segmentation.

### Possible approaches

We can convert pre-trained PyTorch models to ONNX and TensorRT formats, and we can create a ROS 2 package to handle the TensorRT models in Autoware Universe perception to implement them in Autoware.

### Definition of done

- [x] Use pre-trained PyTorch models to detection
- [x] Convert pre-trained PyTorch models to ONNX and TensorRT format
- [x] Deploy TensorRT model with Python
- [x] Deploy TensorRT model with C++
- [x] Create a ROS 2 package in Autoware Universe which use ONNX models
- [ ] Compare RTMDet results with YOLOX segmentation results
- [ ] Decide how can we use the RTMDet segmentation results with image lidar fusion pipeline

**Comment by StepTurtle** [comment]
Here is the results of pre-trained models shared in [this link](https://github.com/open-mmlab/mmdetection/tree/3.x/configs/rtmdet#rtmdet-an-empirical-study-of-designing-real-time-object-detectors) from mmdetection.

**Results:**

| Model  | Score Threshold  | NMS Threshold  | Video Link |
|---|---|---|---|
| RTMDet-Ins-s | 0.3  | 0.3  | [Video Link](https://youtu.be/rMZjsiv6V5o) |
| RTMDet-Ins-x | 0.3  | 0.3  | [Video Link](https://youtu.be/XTCghOU1lDY) |

---

For now, I tested the pre-trained models shared by mmdetection using mmdetection tools. Also they provide a couple of tools to convert `.pth` models to `.onnx` and `.engine` models. I check the both of model results and they are looking same and I think we can say their tools clearly convert the models.

Right now I am trying to handle how can use TensorRT models in `cpp` with TensorRT libraries.

**Comment by StepTurtle** [comment]
I deploy TensorRT engine to Python and I get some consistent results.

 | Model  | Score Threshold  | NMS Threshold  | Detection Time Per Image | Video Link |
|---|---|---|---|---|
| RTMDet-Ins-s (TensorRT) | 0.3  | NO NMS  | ~20 ms | [Video Link](https://youtu.be/ge8leR9OsOs) |
| RTMDet-Ins-x (TensorRT) | 0.3  | NO NMS  | ~36 ms | [Video Link](https://youtu.be/rBS3KIzIzLI) |

> [!WARNING]
> In some parts of the video, you may see incorrect class names. For example, you might see both truck and car class names assigned to a vehicle. This is because I didn't run NMS when I deployed it in Python. I plan to fix this when I deploy it in C++.

---

So, my next plan is doing same things in C++, I'm in deploying engine model in C++

**Comment by StepTurtle** [comment]
I am sharing the results from TensorRT deployment to C++ but right now there is some problems with the results.

I cannot see the exactly same results with the Python deployment. When I check the bounding box and score results I see everything same with Python but when I check the labels, the results are very different. I am currently trying to solve it.

| Model  | Score Threshold  | NMS Threshold  | Detection Time Per Image | Video Link |
|---|---|---|---|---|
| RTMDet-Ins-s (TensorRT) | 0.3  | NO NMS  | ~8 ms | [Video Link](https://youtu.be/sSZ3Gz0L6EQ) |
| RTMDet-Ins-x (TensorRT) | 0.3  | NO NMS  | ~22 ms | [Video Link]() |

---

You can find the scripts I used for deployment at these links:
- Python: https://gist.github.com/StepTurtle/d69ac76d8bd70643fc3899045659bf35
- Cpp: https://gist.github.com/StepTurtle/37aae5afb52fc4d9186296dc5b2fb7c3

**Comment by StepTurtle** [comment]
Fatih suggest to:
- check YOLOX detection time and compare with RTMDet.
- test with datasets and compare scores, bounding boxes and so on with YOLOX.

**Comment by StepTurtle** [comment]
I just started to deploy RTMDet to ROS 2. Here is the repository: https://github.com/leo-drive/tensorrt_rtmdet/

- Right now I have similar results with my previous works.
- Unlike my previous work, it converts ONNX model to TensorRT model on first run and you do not need provide TensorRT model to this package, ONNX is enough.

| Model  | Score Threshold  | NMS Threshold  | Video Link |
|---|---|---|---|
| RTMDet-Ins-s | 0.3  | NO NMS | [Video Link](https://youtu.be/MJpshhxHu8E) |
| RTMDet-Ins-x | 0.3  | NO NMS | [Video Link](https://youtu.be/LFmFXQSM58U) |

---

I plan to complete the porting to ROS2 by performing the following steps:
- RTMDet uses custom TensorRT plugin and It is loading with `dlopen()` right now, find better way.
- For now, preprocessing jobs works on CPU (with OpenCV), I should use CUDA to preprocessing.
- Also post-processing runs on CPU. I will try to convert it CUDA.
- There is no NMS right now, solve it
- There are a lot of hard coded part right now, remove them.

**Comment by StepTurtle** [comment]
✅ **CUDA preprocessing is OK**

Preprocess Time Comparison per Image:

| Device | Average Time (ms) | Standard Deviation (ms) |
|--------|-------------------|------------------------|
| CPU    | `6.87286 ms` | `1.58603 ms` |
| CUDA   | `1.61101 ms` | `0.470295 ms` |

✅ **NMS is added**

When I checked the model, I saw that the existing NMS algorithm within the model was already working. Despite NMS working, the reason we saw overlapping boxes was because the detected objects belonged to different classes. To prevent this, I added a simple NMS algorithm that selects the one with the higher score for overlapping detections.

| Video Link | https://youtu.be/MHPZ930NemE |
|----------------|--------------------|

NMS Time per Image

|  | Average Time (ms) | Standard Deviation (ms) |
|----------------|--------------------|-------------------------|
| NMS | `0.00172242 ms` | `0.000780426 ms` |

🟨 **Plugin Loading**

Autoware has the same logic that I used to load the TensorRT plugin.

```cpp
for (const auto & plugin_path : plugin_paths) {
    int32_t flags{RTLD_LAZY};
    void * handle = dlopen(plugin_path.c_str(), flags); // Plugin path is '/path/to/plugin.so'
    if (!handle) {
      logger_.log(nvinfer1::ILogger::Severity::kERROR, "Could not load plugin library");
    }
}
```

After compilation, a file with the extension '.so' is created. This file stored in `build` and it should be parameter of `dlopen()` function.

Is there any information about can we handle this in Cmake. If we cannot, how can I provide the path to the file located inside the 'build' folder?

I was able to load the plugin using the file paths below:
- `./build/tensorrt_rtmdet/libtensorrt_rtmdet_plugin.so` (relative path from workspace)
- `/home/user/projects/workspace/build/tensorrt_rtmdet/libtensorrt_rtmdet_plugin.so` (absolute path)

🟨 **`int8` Precision Option**

There are three precision option (`fp16`, `fp32` and `int8`) and one of them was not work. In the current situation, it is working, but the result is not entirely correct. Watch video to see problem:

| Video Link | https://youtu.be/3YlY3a9Xnpk |
|----------------|--------------------|

✅ **Hard Coded Parts, Launch and Config files**

I removed hard codded parts and create example launch and config files.

🟨 **Post process and Visualization**

In current implementation visualization and postprocess parts works on CPU. I can't figure out how I can do these on the GPU.

|  | Average Time (ms) | Standard Deviation (ms) |
|----------------|--------------------|-------------------------|
| Preprocess | `1.61101 ms` | `0.470295 ms` |
| Inference | `1.71506 ms` | `0.757723 ms` |
| Postprocess | `13.4203 ms` | `0.628055 ms` |
| Visualization | `23.4338 ms` | `6.98011 ms` |

Following table shows the total time for `preprocess`, `inference` and `postprocess` processes. It don't contain `visualization`

|| Average Time (ms) | Standard Deviation (ms) |
|----------------|--------------------|-------------------------|
| Total | `15.791 ms` | `1.58901 ms` |

- The part which fill the message and publish the result not implement yet. I don't think it will be waste a lot of time.

❓**Message Type for Outputs**

YOLOX semantic segmentation uses following message types:

- For objects: `tier4_perception_msgs::msg::DetectedObjectsWithFeature`
- For semantic mask: 'sensor_msgs::msg::Image`

There is no message defination for instance segmentation, so my plan is creating a new message type which combines the current detection message and new instance segmentation information.

Should I create under [autoware_msgs](https://github.com/autowarefoundation/autoware_msgs) or [tier4_autoware_msgs](https://github.com/tier4/tier4_autoware_msgs/tree/tier4/universe)

---

:memo: An RTX 3090 GPU was used for the time calculations and benchmarking

**Comment by StepTurtle** [comment]
When I ran 8 separate RTMDet nodes, I obtained the results in the table below, numbers represents the process times **per image**:

|        | Average Time (ms) | Standard Deviation (ms) | Min (ms) | Max (ms) |
| ------ | ----------------- | ----------------------- | -------- | -------- |
| node-0 | `43.5821`           | `17.0037`                 | `16.2328`  | `102.521`  |
| node-1 | `42.7286`           | `15.8802`                 | `22.3733`  | `98.7796`  |
| node-2 | `41.4004`           | `15.0862`                 | `15.0169`  | `87.0995`  |
| node-3 | `42.3738`           | `15.7069`                 | `21.1178`  | `105.505`  |
| node-4 | `36.4766`           | `13.2308`                 | `20.3997`  | `81.181`   |
| node-5 | `42.2531`           | `16.1687`                 | `16.265`   | `91.1761`  |
| node-6 | `35.9258`           | `13.5001`                 | `19.7352`  | `76.6217`  |
| node-7 | `36.4776`           | `15.4815`                 | `21.2165`  | `80.9634`  |

**Computer Specifications**

| Device | Model |
|------|------|
| **GPU** | GeForce RTX 3090 (24 gB VRAM) |
| **GPU** | AMD® Ryzen 7 2700x eight-core processor × 16 |
| **Memory** | 32 gB |

**Comment by StepTurtle** [comment]
Following tables shows the process times for `RTMDet` and `YOLOX` model with single and multiple camera configurations

**Computer Specifications**

| Device | Model |
|------|------|
| **GPU** | GeForce RTX 3090 (24 gB VRAM) |
| **GPU** | AMD® Ryzen 7 2700x eight-core processor × 16 |
| **Memory** | 32 gB |

## RTMDet

**Single Camera**

|               | Average Time (ms) | Standard Deviation (ms) |
| ------------- | ----------------- | ----------------------- |
| Preprocess    | `0.488408 ms`     | `0.364109 ms`           |
| Inference     | `2.93087 ms`      | `0.757723 ms`           |
| Postprocess   | `12.0955 ms`      | `0.912801 ms`<br>       |
| Visualization | `24.4686 ms`      | `7.0474 ms`             |
| Fill Mssage   | `8.88453 ms`      | `2.69375 ms`            |
| Total         | `26.1742 ms`      | `5.0135 ms`             |

**Multiple Camera**

|        | Average Time (ms) | Standard Deviation (ms) | Min (ms) | Max (ms) |
| ------ | ----------------- | ----------------------- | -------- | -------- |
| node-0 | 72.119            | 27.7871                 | 23.6683  | 170.952  |
| node-1 | 74.4599           | 28.5641                 | 36.4258  | 189.597  |
| node-2 | 86.9789           | 25.7673                 | 36.9122  | 191.112  |
| node-3 | 68.1571           | 25.7322                 | 32.5606  | 202.151  |
| node-4 | 76.4576           | 24.6437                 | 38.4349  | 147.174  |
| node-5 | 86.094            | 26.974                  | 32.8472  | 184.179  |
| node-6 | 79.927            | 28.3585                 | 39.9874  | 179.886  |
| node-7 | 78.0745           | 29.8809                 | 23.6683  | 162.812  |

## YOLOX

**Single Camera**

|               | Average Time (ms) | Standard Deviation (ms) |
| ------------- | ----------------- | ----------------------- |
| Preprocess    | `0.645836 ms`     | `1.06275 ms`            |
| Inference     | `1.35646 ms`      | `0.860102 ms`           |
| Postprocess   | `4.16757 ms`      | `3.2164 ms`<br>         |
| Visualization | `6.27042 ms`      | `1.73397 ms`            |
| Fill Mssage   | `2.73657 ms`      | `0.390568 ms`           |
| Total         | `10.7695 ms`      | `1.61752 ms`            |

**Multiple Camera**

|        | Average Time (ms) | Standard Deviation (ms) | Min (ms) | Max (ms) |
| ------ | ----------------- | ----------------------- | -------- | -------- |
| node-0 | 65.3484           | 14.5957                 | 33.0228  | 132.028  |
| node-1 | 68.2993           | 16.6715                 | 40.3608  | 133.458  |
| node-2 | 69.3783           | 18.46                   | 33.4685  | 123.697  |
| node-3 | 66.2178           | 17.1871                 | 38.1942  | 130.991  |
| node-4 | 66.7007           | 16.8195                 | 36.598   | 116.047  |
| node-5 | 67.7668           | 17.9907                 | 18.2643  | 160.636  |
| node-6 | 67.4877           | 16.112                  | 36.5241  | 147.61   |
| node-7 | 68.5824           | 17.7861                 | 22.1659  | 126.883  |

**Comment by StepTurtle** [comment]
Discussion on new message type for instance segmentation results: https://github.com/orgs/autowarefoundation/discussions/5047

**Comment by StepTurtle** [comment]
Last updates: https://www.youtube.com/watch?v=N8qrGAxzSJM

**Comment by kminoda** [comment]
@StepTurtle Hi, first of all, thank you for your contribution to Autoware :pray: 

However, I am not sure whether it would be better to merge this into Autoware at this moment (sorry for bringing this up after all of the reviews :pray: )

I have several questions to ask:

1. What is the specific use case that you want to solve with this instance segmentation? Can't it be solved with semantic segmentation e.g. YOLOX in current Autoware?
2. What do you think about making RTMDet (as well as trt_nms_batch) as a separate repository for now, and add that repository to, say, `pilot-auto.leodrive/autoware.repos`? That way, you do not need to go through all the reviewing process, worry about breaking Autoware, care about some latest dependency versions (e.g. [TensorRT](https://github.com/orgs/autowarefoundation/discussions/5300)), etc. We can merge to Autoware once it is clear that this is beneficial to AWF community.

Let me know your thoughts.

**Comment by xmfcx** [comment]
<img src="https://github.com/user-attachments/assets/92cce229-f67c-472c-8f5c-aa9d37da3a9c" width="500">

It also has competitive speed.

<img src="https://github.com/user-attachments/assets/84af7485-7a55-49f9-a748-b7eb92ff59a5" width="200">

Why project boxes when you can project instance segmentation masks?

## ❌ Poor bounding box to point cloud performance

`only_allow_inside_cluster = false`

https://github.com/user-attachments/assets/63fbb81d-5ae5-4d28-8918-1763789afa81

ROI cluster fusion code has very complicated rules:
- https://github.com/autowarefoundation/autoware.universe/blob/main/perception/autoware_image_projection_based_fusion/src/roi_cluster_fusion/node.cpp

### ❌ Reduced performance with `only_allow_inside_cluster`

`only_allow_inside_cluster = true`

Probably to solve this issue, you've added `only_allow_inside_cluster` parameter recently. But even that, lead to mis-labeling of wrong objects. And even miss the pedestrians in front. (Green is prediction)

https://github.com/user-attachments/assets/c9f966d2-60b6-49b3-bcb4-88ae998be4a2

## ✅ Project instance segmentation masks

https://github.com/user-attachments/assets/f450af0c-a6af-49d0-9b78-1b089f65f577

Projection code is much more simpler and works all the time.

No need for weird workarounds.

### Why not semantic segmentation?

> Can't it be solved with semantic segmentation e.g. YOLOX in current Autoware?

Why would we use the outdated semantic segmentation technology when we could achieve instance segmentation with similar performance?

![image](https://github.com/user-attachments/assets/3f9fe40b-2e8e-4271-87f0-2e385f6233d0)

https://blog.roboflow.com/difference-semantic-segmentation-instance-segmentation/

![image](https://github.com/user-attachments/assets/baa961f5-870d-4fbd-8a57-fbe540e717d0)

You cannot achieve this level of granularity with semantic segmentation.

**Comment by xmfcx** [comment]
> That way, you do not need to go through all the reviewing process, worry about breaking Autoware, care about some dependency version change (e.g. [TensorRT](https://github.com/orgs/autowarefoundation/discussions/5300)), etc. We can merge to Autoware once it is clear that this is beneficial to AWF community.

I think this is a very backwards way of thinking. The universe is supposed to be for community contributions.

A lot of effort is already been put into this integration. And the author was very cooperative during the process.

The benefits are very clear and I don't see the reason for refusing to accept a feature like this.

It doesn't even affect the existing repositories. I am very confused about your proposal.

**Comment by armaganarsln** [comment]
> @StepTurtle Hi, first of all, thank you for your contribution to Autoware 🙏
> 
> However, I am not sure whether it would be better to merge this into Autoware at this moment (sorry for bringing this up after all of the reviews 🙏 )
> 
> I have several questions to ask:
> 
> 1. What is the specific use case that you want to solve with this instance segmentation? Can't it be solved with semantic segmentation e.g. YOLOX in current Autoware?
> 2. What do you think about making RTMDet (as well as trt_nms_batch) as a separate repository for now, and add that repository to, say, `pilot-auto.leodrive/autoware.repos`? That way, you do not need to go through all the reviewing process, worry about breaking Autoware, care about some dependency version change (e.g. [TensorRT](https://github.com/orgs/autowarefoundation/discussions/5300)), etc. We can merge to Autoware once it is clear that this is beneficial to AWF community.
> 
> Let me know your thoughts.

This is too late to suggest. All work has been done and it has to be merged to the main now. I am sorry for not giving any other options but these comments had to be made earlier.

**Comment by armaganarsln** [comment]
@kminoda san, thank you and I am sorry for misunderstanding you on your approach. I thought you don't want it to be used or be in the main. Anyway there is still a lot work to be done for it to be useful so let's focus on that right now and see if there will be an improvement in the end in the overall false detections.
Thank you for your support.

**Comment by StepTurtle** [comment]
Hey @kminoda , could you update me on your latest decision? Are the reviews ongoing, or are there other concerns you're currently considering? It would be helpful to know the latest status so I can continue working.

**Comment by StepTurtle** [comment]
Closing the discussion and PRs as `won't do` due to a lack of motivation, requests on this topic, and no feedback received for a while.

**Comment by armaganarsln** [comment]
@xmfcx @mitsudome-r could you please check why there are no reviews on this?

**Comment by kminoda** [comment]
@StepTurtle (cc: @mitsudome-r @xmfcx ) I am sorry I overlooked the notification and also failed to assign reviewers properly. We have been obsessed with some internal development items recently. 

Please kindly open the PR again if you are willing to merge. If so, let us assign reviewers again. Thank you for your understanding and patience.

**Comment by mitsudome-r** [comment]
@armaganarsln @StepTurtle 
I'm sorry about the late reviews. 

We've discussed this in Sensing & Perception WG today, and as @kminoda -san mentioned, the reviewers were busy for the past few months with some projects and wasn't able to spend enough time on it.

That being said, I will raise this in TSC and SPC we definitely need improvements in our review pipeline to speed up the process. It doesn't really make sense to have TIER IV "blocking" the PRs for packages that is not owned by TIER IV. We have set the rule that if a PR doesn't have reaction for over a week, either @xmfcx and @mitsudome-r will pick them up, but we probably need more reviewers to support us.

**Comment by armaganarsln** [comment]
Thank you both for your replies, @kminoda and @mitsudome-r.

I believe this is an important topic that we should discuss further at both TSC and SPC.

I have a straightforward and practical solution to address this issue and similar problems, but it will require being open and making changes. My proposed solution is simple: tasks should be created, assigned, and organized by reviewers/decision-makers, and no one should be able to create large tasks for AV Software (besides bugfix etc.), such as this one, and then ask for a review.

From my observation, similar issues created by others in the past have often ended in failure (including AWSIM Labs). If managers are willing to focus on creating and organizing tasks rather than directly developing them or saying no to other proposals, we could eliminate this inefficiency.

Specifically, my expectation would be for @kminoda-san not to focus on developing but rather on creating tasks, issues, and reviewing or assigning people to review them. If this approach is not adopted, I fear that meaningful and impactful contributions to AV software will remain limited to TIER IV.

In the past, I had hoped that @xmfcx and @mitsudome-r could take on this role, and this was the idea behind Autoware Labs—fixing inefficiencies through direct task management—but unfortunately, it didn’t work out because in the end the reviewer is someone else.

To summarize, my proposal is to encourage decision-makers to take a more active role in managing and determining what they want to be developed in Autoware not their own repository. If individuals like @kminoda, @TakaHoribe, and other managers in perception, planning, etc., are willing to create and review issues, I believe we can achieve success.

**Comment by kminoda** [comment]
@StepTurtle Hello, thank you for the summary, I understood the situation. Let us know once you reopen the PR :pray:

**Comment by StepTurtle** [comment]
@kminoda OK.

Old ones reopened, i will create for PR for `mask_cluster_fusion` within this week.

**Comment by amadeuszsz** [comment]
@StepTurtle 
Thank you for reopening RTMDet PRs. I will support review process, please give me some time, but not more than a week

**Comment by mitsudome-r** [comment]
I had a discussion with @StepTurtle on this to clarify who would be working on each task explained [here](https://github.com/autowarefoundation/autoware.universe/issues/7235#issuecomment-2482619675).

There are currently two one PRs:
* https://github.com/autowarefoundation/autoware.universe/pull/8165
  * will be reviewed by TIER IV Perception team (@amadeuszsz )
* https://github.com/autowarefoundation/autoware_internal_msgs/pull/23
  * reviewed @mitsudome-r 

We need the following PRs to be created:

- [ ] Update PR to roi_cluster fusion @StepTurtle 
- [ ] Update autoware.repos file to add https://github.com/autowarefoundation/trt_batched_nms @StepTurtle 
- [ ] Upload onnx files to [google drive](https://drive.google.com/drive/folders/1P1JyMrstJIN9Y4F947nUkbKxQ2JDntgF) @StepTurtle 
- [ ] Copy the uploaded onnx/label files to AWF's AWS S3 bucket server so that we can [download it as an artifact](https://github.com/autowarefoundation/autoware/blob/main/ansible/roles/artifacts/tasks/main.yaml) @mitsudome-r

**Comment by StepTurtle** [comment]
@mitsudome-r, I update the items from the TODO list in your message and let me know if there's anything else.

**Comment by StepTurtle** [comment]
@mitsudome-r Reviews seem to be nearly complete. Did you upload the artifacts?

**Comment by mitsudome-r** [comment]
@StepTurtle Thanks for the reminder. I have uploaded the artifacts here: https://autoware-files.s3.us-west-2.amazonaws.com/models/tensorrt_rtmdet_onnx_models.tar.gz

**Comment by StepTurtle** [comment]
@mitsudome-r I have following error when I tried to download the data:

```bash
> wget -P /tmp/ https://autoware-files.s3.us-west-2.amazonaws.com/models/tensorrt_rtmdet_onnx_models.tar.gz

--2024-12-13 11:56:35--  https://autoware-files.s3.us-west-2.amazonaws.com/models/tensorrt_rtmdet_onnx_models.tar.gz
Resolving autoware-files.s3.us-west-2.amazonaws.com (autoware-files.s3.us-west-2.amazonaws.com)... 52.92.194.34, 52.92.229.170, 52.218.216.57, ...
Connecting to autoware-files.s3.us-west-2.amazonaws.com (autoware-files.s3.us-west-2.amazonaws.com)|52.92.194.34|:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
2024-12-13 11:56:36 ERROR 403: Forbidden.
```

**Comment by StepTurtle** [comment]
@mitsudome-r ping

**Comment by xmfcx** [comment]
It was missing a public read permission, fixed it @mitsudome-r @StepTurtle the link should work for everyone now.

**Comment by mitsudome-r** [comment]
@xmfcx Thanks!

**Comment by xmfcx** [comment]
## Updates

cc. @mitsudome-r @StepTurtle @amadeuszsz @YoshiRi @kminoda 

### Merged

- https://github.com/autowarefoundation/autoware_internal_msgs/pull/29
- https://github.com/autowarefoundation/autoware/pull/5493
- https://github.com/autowarefoundation/autoware/pull/5605

### Remaining

- https://github.com/autowarefoundation/autoware.universe/pull/8165
  - Requires update due to breaking change https://github.com/autowarefoundation/autoware/issues/5607
- https://github.com/autowarefoundation/autoware.universe/pull/9482
  - Was not reviewed yet, requires rebase conflict resolution.
- For both PRs, a ros2 bag file with detailed steps to reproduce would be useful for reviewing.

**Comment by VeeranjaneyuluToka** [comment]
@StepTurtle , can i use the model trained using mmyolo - yolox and converted .onnx file using mmdepoly in autoware-perception tensorrt-yolox node directly as they use different github to train yolox model as per autoware documentation?

**Comment by xmfcx** [comment]
> @StepTurtle , can i use the model trained using mmyolo - yolox and converted .onnx file using mmdepoly in autoware-perception tensorrt-yolox node directly as they use different github to train yolox model as per autoware documentation?

@VeeranjaneyuluToka apologies for the very late reply, this thread went quiet and your question deserved an answer.

**Short version: not directly.** `autoware_tensorrt_yolox` does not care which repo trained the weights, but it does care about the shape of the exported ONNX graph, and a stock mmdeploy export will not work as-is.

The node picks its post-processing purely from the number of TensorRT I/O tensors ([`tensorrt_yolox.cpp:131`](https://github.com/autowarefoundation/autoware_universe/blob/3194c4d2cbd9aa9acbaf116cd5d5e865fc39b646/perception/autoware_tensorrt_yolox/src/tensorrt_yolox.cpp#L131)):

- **2 tensors**: plain YOLOX head, the node decodes the raw output itself.
- **5 tensors**: the graph ends in `EfficientNMS_TRT`, so the node reads the decoded boxes directly.
- **anything else**: falls into `default`, which sets `multitask_++` and treats the extra output as a **semantic segmentation head**.

A stock mmdeploy end-to-end export gives `dets` plus `labels`, so 3 tensors, and it lands in that last branch. There is no graph validation anywhere, so instead of a clean error you get silently wrong output.

Two ways to make it work:

1. Re-export so the graph ends in `EfficientNMS_TRT` (5 outputs, boxes in the letterboxed network-input pixel space), or
2. export the raw head in the Megvii layout so it hits the 2-tensor path.

Either way, match the node's preprocessing, because none of it is configurable: it letterboxes with `scale = min(net_w / img_w, net_h / img_h)` ([`tensorrt_yolox.cpp:285`](https://github.com/autowarefoundation/autoware_universe/blob/3194c4d2cbd9aa9acbaf116cd5d5e865fc39b646/perception/autoware_tensorrt_yolox/src/tensorrt_yolox.cpp#L285)), pads with **114** ([`preprocess.cu:109`](https://github.com/autowarefoundation/autoware_universe/blob/3194c4d2cbd9aa9acbaf116cd5d5e865fc39b646/perception/autoware_tensorrt_yolox/src/preprocess.cu#L109)), feeds **BGR**, and applies no mean/std normalisation (`norm_factor` is hard-coded to `1.0`). There is also no `num_class` parameter: the class count comes from the number of entries in the file you pass as `label_path`. Note the package README still calls that parameter `label_file` in a couple of places, which is stale; the actual ROS parameter is `label_path`.

**One thing specific to this thread:** the RTMDet work is not going forward and is not supported. `autoware_tensorrt_rtmdet` was never merged (#8165 is closed), this issue is being closed, and we are removing the leftover model download and the `trt_batched_nms` dependency. `autoware_tensorrt_yolox` is the supported 2D camera detector. I will follow up here shortly with the full rationale and a link to the cleanup tracking issue.

If the export still gives you trouble, please open a fresh issue on `autoware_universe` and we will follow up there rather than on a closed thread.

**Comment by xmfcx** [comment]
Picking this thread back up to close it out properly, since it has been quiet for a while.

**Decision: we are not taking RTMDet forward.** The mechanical removal is tracked in autowarefoundation/autoware#7252.

Why, in short. #8165 (`autoware_tensorrt_rtmdet`) was reviewed thoroughly, 86 commits and 50 review threads all resolved, and was approved by @amadeuszsz on 2024-12-12. It ended up approved, not rejected on code quality. It stalled because @StepTurtle stepped away in early 2025 and nobody took it over, myself included. I closed it on 2025-12-26 after getting it to build against current Autoware and then hitting a TensorRT 10 ONNX parse failure.

Three things kept it from being worth restarting:

- **No consumer.** Nothing in `autoware_universe` or `autoware_core` produces or consumes 2D instance masks. `SegmentationMask.msg` has been merged for 20 months with zero producers and zero consumers, and the one consumer candidate, #9482, never got a second review from us.
- **No benchmark.** The definition of done says "Compare RTMDet results with YOLOX segmentation results", and that has been unticked for 26 months. Latency was posted several times; the accuracy comparison was never produced, by the author or by any of us reviewing it. So we still have no evidence it improves anything measurable.
- **Unmaintained upstream.** mmdetection's last commit on `main` is 2024-02-05 and mmdeploy's is 2024-09-30, so we would be shipping a 2024 ONNX we have demonstrably failed to re-export, plus a forked NVIDIA NMS CUDA plugin no other Autoware package needs.

For the record, the TensorRT 10 blocker itself looked fixable. The `/Add_2` Int32 vs Int64 parse error is a bounded defect in `autowarefoundation/trt_batched_nms`, which hardcodes `kINT32` in `getOutputDataType()` and `supportsFormatCombination()` and ships only the `int *` path in `src/nms/gatherNMSOutputs.cu`; `VBTI-development/onedl-mmdeploy` guards all three with `#if NV_TENSORRT_MAJOR < 10` under Apache-2.0. Nobody has carried that through end to end on Autoware's own TensorRT, so it is a documented path rather than a demonstrated one. It is not the reason we are stopping.

What is being removed: the `tensorrt_rtmdet` model download (462 MB fetched per install, about 957 MB on disk) and the `universe/external/trt_batched_nms` entry in `repositories/autoware.repos`. What is being kept: `SegmentationMask.msg`, marked deprecated rather than removed so any out-of-tree consumer keeps building, and #9018 (`enqueueV3` support in `autoware_tensorrt_common`), which came out of this effort, was merged, and is used today by `autoware_camera_streampetr`, `autoware_tensorrt_vad` and `autoware_diffusion_planner`.

The capability itself is still worth having, and I would rather that be on the record than buried. `roi_cluster_fusion` still associates by winner-take-all box IoU, and it picked up two more refinements in 2026 (#11828 per-class IoU thresholds, #12019 a pedestrian size gate stated as reducing false positives), so association quality in that path is clearly still being worked on. If anyone wants to revisit it, the preconditions are an owner with a maintainer standing behind them, a benchmark against the current `roi_cluster_fusion` baseline rather than 2024's, and a consumer that lands in the same cycle as the producer. I would also look at a model that exports plain ONNX with no custom TensorRT plugin, since that plugin is exactly what killed this attempt.

Thank you @StepTurtle for the work and for documenting it so thoroughly as you went. The write-ups in this thread are why a useful post-mortem was possible at all.

Closing this issue now that the decision is recorded. The mechanical removal is tracked to completion in autowarefoundation/autoware#7252, so there is no reason to keep this one open on top of it. #9482 will be closed separately, with porting notes preserved, since that node is model-agnostic and is worth reviving if an instance-mask producer ever shows up.
